Guides/Unique.md

```diff
@@ -14,29 +14,36 @@ element.
 let numbers = [1, 2, 3, 3, 2, 3, 3, 2, 2, 2, 1]
 
 let unique = numbers.uniqued()
-// unique == [1, 2, 3]
+// Array(unique) == [1, 2, 3]
 ```
 
 ## Detailed Design
 
 Both methods are available for sequences, with the simplest limited to
 when the element type conforms to `Hashable`. Both methods preserve
-the relative order of the elements.
+the relative order of the elements. `uniqued(on:)` has a matching lazy
+version that is added to `LazySequenceProtocol`.
 
 ```swift
 extension Sequence where Element: Hashable {
-    func uniqued() -> [Element]
+    func uniqued() -> Uniqued<Self, Element>
 }
 
 extension Sequence {
-    func uniqued<T>(on: (Element) throws -> T) rethrows -> [Element]
-        where T: Hashable
+    func uniqued<Subject>(on projection: (Element) throws -> Subject) rethrows -> [Element]
+        where Subject: Hashable
+}
+
+extension LazySequenceProtocol {
+    func uniqued<Subject>(on projection: @escaping (Element) -> Subject) -> Uniqued<Self, Subject>
+        where Subject: Hashable
 }
 ```
 
 ### Complexity
 
-The `uniqued` methods are O(_n_) in both time and space complexity.
+The eager `uniqued(on:)` method is O(_n_) in both time and space complexity.
+The lazy versions are O(_1_).
```

Comment:
I think it's a very salient consideration mentioned by @kylemacomber that most uses seem to expect an `Array`, in which case a `.lazy.uniqued` design would make more sense.

Additionally, this discrepancy here, where `uniqued(on:)` isn't lazy by default but `uniqued` would be, seems like it invites confusion. I certainly would not expect that supplying a custom predicate would change the complexity or behavior so fundamentally, and I can see that issue arising when people make changes to their code and encounter this difference. Therefore, if it makes sense to make `uniqued` lazy, then I think the same change should be applied to `uniqued(on:)`.

Reply:
Do people expect `uniqued()` to return an array, or do implementations usually simply return an array because it's easier to implement that way? I genuinely don't know the answer to this.

I share your concern about the discrepancy between `uniqued()` and `uniqued(on:)`, but I do think that viewing them in isolation, this change is consistent with other sequence operations in the standard library. Operations that can be lazy without having to compromise (except perhaps on the return type) typically are, even when called on a sequence not conforming to `LazySequenceProtocol`. `Collection`'s `joined()` and `reversed()` fall into this category, and I argue that `uniqued()` does as well. `uniqued(on:)` does not, because making it lazy would require the closure to be escaping and non-throwing.

Reply:
Do any other algorithms in this library or the standard library differ in laziness depending on the presence of a custom predicate?

Reply:
I believe the pair of operations that comes closest is `joined()` being lazy and `flatMap { $0 }` being eager while effectively doing the same thing. Of course they don't have similar names like `uniqued()` and `uniqued(on:)` do.

This is a fairly unique situation because other potential pairs lack one of the variants. `chunked(by: ==)` and `compactMap { $0 }` have no corresponding closure-less version, and operations like `zip`, `reversed`, and `combinations` have no versions that do take a closure.

Reply:
Ya I think the `joined()`/`flatMap { $0 }` and `compacted()`/`compactMap { $0 }` (see #112) serve as good precedent of similar of lazy/eager pairs.

To try to articulate the philosophy:

1. Anything that can efficiently be lazy should be lazy, because (i) it can avoid extra work (e.g. unnecessary computation or allocation) and (ii) it's easy to go from lazy to eager by constructing an Array, but it's impossible to go the other way.
2. We require an explicit `.lazy` for algorithms that take a closure to emphasize that (i) the closure will not run immediately and (ii) may run more than once per element in the collection, which can introduce a surprising vector for error.
3. An algorithm shouldn't be lazy if its being lazy would pessimize its runtime complexity. For example, a lazy `reversed` adapter over a plain `Collection` would be absurd because looping over all of its indices would be O(n²).

Reply:
Ya I think the `joined()`/`flatMap { $0 }` and `compacted()`/`compactMap { $0 }` (see #112) serve as good precedent of similar of lazy/eager pairs.

To try to articulate the philosophy:

1. Anything that can efficiently be lazy should be lazy, because (i) it can avoid extra work (e.g. unnecessary computation or allocation) and (ii) it's easy to go from lazy to eager by constructing an Array, but it's impossible to go the other way.
2. Require an explicit `.lazy` for algorithms that take a closure to emphasize that (i) the closure will not run immediately and (ii) may run more than once per element in the collection, which can introduce a surprising vector for error.
3. An algorithm shouldn't be lazy if its being lazy would pessimize its runtime complexity. For example, a lazy `reversed` adapter over a plain `Collection` would be absurd because looping over all of its indices would be O(n²).

Reply:
Now that we have `OrderedSet` via the Swift Collections packages, I think it's even clearer to me that `uniqued` should be lazy... the ability to do partial computation is really the only thing (other than method call syntax) distinguishing this algorithm from just creating an `OrderedSet`

Reply:
> Now that we have `OrderedSet` via the Swift Collections packages, I think it's even clearer to me that `uniqued` should be lazy... the ability to do partial computation is really the only thing (other than method call syntax) distinguishing this algorithm from just creating an `OrderedSet`

+1